crates/viewer/re_ui/src/notifications.rs

```diff
@@ -0,0 +1,393 @@
+
+use crate::icons;
+use crate::UiExt;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum NotificationLevel {
+    Info,
+    Warning,
+    Error,
+    Success,
+}
+
+impl From<re_log::Level> for NotificationLevel {
+    fn from(value: re_log::Level) -> Self {
+        match value {
+            re_log::Level::Trace | re_log::Level::Debug | re_log::Level::Info => Self::Info,
+            re_log::Level::Warn => Self::Warning,
+            re_log::Level::Error => Self::Error,
+        }
+    }
+}
+
+fn is_relevant(level: re_log::Level) -> bool {
+    matches!(
+        level,
+        re_log::Level::Warn | re_log::Level::Error | re_log::Level::Info
+    )
+}
+
+pub fn notification_toggle_button(
+    ui: &mut egui::Ui,
+    show_notification_panel: &mut bool,
+    has_unread_notifications: bool,
+) {
+    let response = ui.medium_icon_toggle_button(&icons::NOTIFICATION, show_notification_panel);
+
+    if has_unread_notifications {
+        let mut pos = response.rect.right_top();
+        pos.x -= 2.0;
+        pos.y += 2.0;
+        ui.painter().circle_filled(pos, 3.0, hex_color!("#ab0037"));
+    }
+}
+
+struct Notification {
+    level: NotificationLevel,
+    text: String,
+
+    created_at: Zoned,
+    ttl: Duration,
+}
+
+pub struct NotificationUi {
+    /// State of every notification.
+    ///
+    /// Notifications are stored in order of ascending age, so the latest one is at the end.
+    data: Vec<Notification>,
+
+    has_unread_notifications: bool,
+
+    /// Panel that shows all notifications.
+    panel: NotificationPanel,
+
+    /// Toasts that show up for a short time.
+    toasts: Toasts,
+}
+
+impl Default for NotificationUi {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NotificationUi {
+    pub fn new() -> Self {
+        Self {
+            data: Vec::new(),
+            has_unread_notifications: false,
+            panel: NotificationPanel::new(),
+            toasts: Toasts::new(),
+        }
+    }
+
+    pub fn has_unread_notifications(&self) -> bool {
+        self.has_unread_notifications
+    }
+
+    pub fn add(&mut self, level: Level, text: impl Into<String>) {
+        if !is_relevant(level) {
+            return;
+        }
+
+        self.data.push(Notification {
+            level: level.into(),
+            text: text.into(),
+
+            created_at: Zoned::now(),
+            ttl: base_ttl(),
+        });
+        self.has_unread_notifications = true;
+    }
+
+    pub fn success(&mut self, text: impl Into<String>) {
+        self.data.push(Notification {
+            level: NotificationLevel::Success,
+            text: text.into(),
+
+            created_at: Zoned::now(),
+            ttl: base_ttl(),
+        });
+    }
+
+    pub fn ui(&mut self, egui_ctx: &egui::Context, is_panel_visible: &mut bool) {
+        if *is_panel_visible {
+            let escape_pressed =
+                egui_ctx.input_mut(|i| i.consume_key(egui::Modifiers::NONE, egui::Key::Escape));
```

Comment:
From cursory testing,  at least on web, the escape key is not "consumed" enough to avoid loosing your current selection. I'm unsure why this is the case though. But I sure dislike this behaviour, so would be great to find a fix 🙏🏻

Reply:
That should depend only on the order in which selection checks for `Key::Escape` and this code is run, so maybe just need to move some code around